### ISSUE TITLE
recursive from_msgpack implementation will stack overflow

### ISSUE THREAD

**Issue by xgzeng** [nlohmann/json#537]
`from_msgpack` function(etc) parses data recursively, which will cause stack overflow when process deep nested structure. 
It will be DOS vulnerabilities, if we use library to process data from outside.

Following is example code to generate malicious data. `msgpack_pack_array` is not shown here.
```cpp
for (int i = 0; i < 10000; ++i) {
    msgpack_pack_array(buffer, 1);
}
msgpack_pack_array(buffer, 0);
```

**Comment by nlohmann** [comment]
The same issue was detected for CBOR in #517.

**Comment by xgzeng** [comment]
Should we just add a defaut parameter to restrict nesting level?
I think it suits most use case.

**Comment by nlohmann** [comment]
That would be an effective, yet ugly solution. I would rather have a non-recursive version that uses a stack internally so we can enforce a limit there.

**Comment by nlohmann** [comment]
Closed in favor of #517.
